serve/mlc_serve/model/paged_cache_model.py

```diff
@@ -105,14 +105,18 @@ def _is_safe_to_sample(prob_like):
             do_top_p |= top_ps[-1] < 1.0
             do_top_k |= top_ks[-1] != vocab_size
 
-            if not param.presence_penalty == 0.0 or not param.frequency_penalty == 0:
-                for token_id, token_freq in freq.items():
-                    logits[i][token_id] -= token_freq * param.frequency_penalty + param.presence_penalty
+            if not param.presence_penalty == 0.0 or not param.frequency_penalty == 0 and bool(freq):
+                freq_tensor = np.array(list(freq.items()))
+                index = torch.from_numpy(freq_tensor[..., 0]).to(device=logits.device)
+                src = torch.from_numpy(freq_tensor[..., 1]).type_as(logits).to(device=logits.device)
```

Comment:
Use `freq.keys()` and `freq.values()` to create `index` and `src` respectively. No need for `freq_tensor`.

Reply:
updated